src/clr-angular/forms/datepicker/providers/date-io.service.ts

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import {Injectable} from "@angular/core";
+import {Observable} from "rxjs/Observable";
+import {Subject} from "rxjs/Subject";
+
+import {
+    BIG_ENDIAN,
+    DEFAULT_LOCALE_FORMAT,
+    InputDateDisplayFormat,
+    LITTLE_ENDIAN,
+    LITTLE_ENDIAN_REGEX,
+    MIDDLE_ENDIAN,
+    MIDDLE_ENDIAN_REGEX,
+    NON_NEGATIVE_NUMBER,
+    SEPARATORS
+} from "../utils/constants";
+import {getNumberOfDaysInTheMonth, parseToFourDigitYear} from "../utils/date-utils";
+
+import {LocaleHelperService} from "./locale-helper.service";
+
+@Injectable()
+export class DateIOService {
+    public cldrLocaleDateFormat: string = DEFAULT_LOCALE_FORMAT;
+    private localeDisplayFormat: InputDateDisplayFormat = LITTLE_ENDIAN;
+
+    constructor(private _localeHelperService: LocaleHelperService) {
+        this.cldrLocaleDateFormat = this._localeHelperService.localeDateFormat;
+        this.initializeLocaleDisplayFormat();
+    }
+
+    private _inputDate: string;
+
+    get inputDate(): string {
+        return this._inputDate;
+    }
+
+    /**
+     * This setter is used for setting the inputDate when the user types in the input.
+     */
+    set inputDate(value: string) {
+        const date: Date = this.isValidInput(value);
+        if (date) {
+            this.setDate(date);
+            this._inputDate = value;
+        } else {
+            this._inputDate = "";
+            this.setDate(null);
+        }
+    }
+
+    /**
+     * Function to update the date and emit it to the dateChanged subscribers.
+     * This method is used when the user selects the date from the calendar.
+     */
+    updateDate(date: Date) {
+        this.inputDate = this.toLocaleDisplayFormatString(date);
+        this._dateStrUpdated.next(this.inputDate);
+    }
+
+    private _date: Date;
```

Comment:
So that's what I was saying. I'm not sure why you need to store the date and the string version here. This service should essentially be `toString(date: Date)` and `toDate(dateStr: string)`, just parsing/formatting and returning the result.